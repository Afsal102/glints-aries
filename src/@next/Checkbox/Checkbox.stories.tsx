import React from 'react';
import { Story, Meta } from '@storybook/react';

import { Checkbox, CheckboxProps } from './Checkbox';
import { BaseContainer } from '../../Layout/GlintsContainer/GlintsContainer';

(Checkbox as React.FunctionComponent<CheckboxProps>).displayName = 'Checkbox';

export default {
  title: '@next/Checkbox',
  component: Checkbox,
  decorators: [Story => <BaseContainer>{Story()}</BaseContainer>],
} as Meta;

const Template: Story<CheckboxProps> = args => <Checkbox {...args} />;

export const Interactive = Template.bind({});
Interactive.args = { label: 'Label' };
