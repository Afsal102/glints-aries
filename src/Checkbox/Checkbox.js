/* @flow */

import React from 'react';
import { Container } from '../Style/CheckboxStyle';

const Checkbox = (props: Props) => {
  const {
    id,
    label,
    className,
    ...defaultProps
  } = props;

  return (
    <Container className={className}>
      <input type="checkbox" id={id} {...defaultProps} />
      <label htmlFor={id}>
        {label}
      </label>
    </Container>
  );
};

type Props = {
  id: string,
  label: string,
  className: string,
}

export default Checkbox;
