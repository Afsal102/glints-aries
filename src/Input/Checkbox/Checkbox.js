/* @flow */

import React, { Component } from 'react';
import { CheckboxContainer } from '../../Style/Input/CheckboxStyle';

class Checkbox extends Component <Props, State> {
  state = {
    checked: false,
  }

  handleClick = (onClick) => {
    const listener = (e) => {
      const { checked } = this.state;
      this.setState({
        checked: !checked,
      });

      if (onClick !== undefined) {
        return onClick(e);
      }
    };

    return listener;
  }

  render() {
    const {
      id,
      value,
      onClick,
      className,
      ...defaultProps
    } = this.props;

    const { checked } = this.state;

    return (
      <CheckboxContainer
        className={className}
        role="checkbox"
        aria-labelledby={id}
        aria-checked={checked}
        tabIndex="0"
      >
        <input
          type="checkbox"
          id={id}
          value={value}
          onClick={this.handleClick(onClick)}
          {...defaultProps}
        />
        <label
          htmlFor={id}
          tabIndex="-1"
        >
          {value}
        </label>
      </CheckboxContainer>
    );
  }
}

type Props = {
  id: string,
  value: string,
  className: string,
  onClick: Function,
}

type State = {
  checked: boolean,
}

export default Checkbox;
