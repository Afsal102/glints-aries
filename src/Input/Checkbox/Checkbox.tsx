import * as React from 'react';
import classNames from 'classnames';

import { CheckboxContainer } from '../../Style/Input/CheckboxStyle';

const Checkbox: React.FunctionComponent<CheckboxProps> = ({
  id,
  label,
  value,
  onClick,
  size = 'small',
  border = false,
  className,
  ...restProps
}: CheckboxProps) => {
  const [checked, setChecked] = React.useState(false);

  const handleClick = (e: React.MouseEvent<HTMLInputElement, MouseEvent>) => {
    setChecked(checked => !checked);
    if (onClick !== undefined) {
      return onClick(e);
    }
  };

  return (
    <CheckboxContainer
      className={classNames('aries-checkbox', { border, checked }, className)}
      role="checkbox"
      aria-labelledby={id}
      aria-checked={checked}
      tabIndex={0}
      size={size}
    >
      <input
        type="checkbox"
        id={id}
        value={value}
        onClick={handleClick}
        className={classNames({ border })}
        {...restProps}
      />
      <label htmlFor={id} tabIndex={-1}>
        {label || value}
      </label>
    </CheckboxContainer>
  );
};

type HTMLInputProps = Omit<React.HTMLProps<HTMLInputElement>, 'size' | 'label'>;

export interface CheckboxProps extends HTMLInputProps {
  label?: React.ReactNode;
  size?: 'large' | 'small';
  border?: boolean;
}

export default Checkbox;
