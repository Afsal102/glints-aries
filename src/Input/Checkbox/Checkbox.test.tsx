import * as React from 'react';

import * as renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { fireEvent, render } from '@testing-library/react';

import Checkbox, { CheckboxProps } from './Checkbox';

const props = {
  id: 'software-engineer',
  value: 'Software Engineer',
  onClick: jest.fn().mockImplementation(event => event.target.value),
};

function setupCheckbox(props: CheckboxProps) {
  const { id, value, onClick, ...restProps } = props;
  const { getByText, getByLabelText, ...utils } = render(
    <Checkbox id={id} value={value} onClick={onClick} {...restProps} />
  );
  const checkboxInput = getByLabelText(value as string) as HTMLInputElement;
  const checkboxLabel = getByText(value as string) as HTMLLabelElement;
  return { checkboxInput, checkboxLabel, utils };
}

afterEach(() => {
  props.onClick.mockClear();
});

it(`<Checkbox> should render an input with id, value and onClick props and a label with the text ${props.value}`, () => {
  const CheckboxSnapshot = renderer
    .create(
      <Checkbox id={props.id} value={props.value} onClick={props.onClick} />
    )
    .toJSON();
  expect(CheckboxSnapshot).toMatchSnapshot();
});

it('when toggling checkbox, it should fire onClick once and become checked, then fire onClick once and become unchecked', () => {
  const { checkboxInput, checkboxLabel } = setupCheckbox(props);

  fireEvent.click(checkboxLabel);
  expect(props.onClick).toHaveBeenCalledTimes(1);
  expect(checkboxInput.checked).toEqual(true);

  props.onClick.mockClear();

  fireEvent.click(checkboxLabel);
  expect(props.onClick).toHaveBeenCalledTimes(1);
  expect(checkboxInput.checked).toEqual(false);
});

it('when passed label, it should render the label', () => {
  const label = <p data-testid="label-child">label child</p>;
  const { getByTestId } = render(
    <Checkbox id={props.id} value={props.value} label={label} />
  );
  const LabelChild = getByTestId('label-child');
  expect(document.querySelector('label')).toContainElement(LabelChild);
});

describe('when it is rendered', () => {
  it('should display the correct label', () => {
    const { checkboxInput } = setupCheckbox(props);
    expect(checkboxInput).toBeTruthy();
  });

  it('should display an unchecked checkbox', () => {
    const { checkboxInput } = setupCheckbox(props);
    expect(checkboxInput.checked).toEqual(false);
  });

  it('should have the correct input id and value', async () => {
    const { checkboxInput } = setupCheckbox(props);
    expect(checkboxInput.id).toEqual(props.id);
    expect(checkboxInput.value).toEqual(props.value);
  });

  it('should return the input value when onClick is passed a function: event => event.target.value', () => {
    const { checkboxLabel } = setupCheckbox(props);
    fireEvent.click(checkboxLabel);
    expect(props.onClick.mock.results[0].value).toEqual(props.value);
  });
});
