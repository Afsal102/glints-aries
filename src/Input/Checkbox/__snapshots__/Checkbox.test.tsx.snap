// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox> should render an input with id, value and onClick props and a label with the text Software Engineer 1`] = `
<div
  aria-checked={false}
  aria-labelledby="software-engineer"
  className="CheckboxStyle__CheckboxContainer-lmispl-0 jekiyq aries-checkbox"
  role="checkbox"
  size="small"
  tabIndex={0}
>
  <input
    className=""
    id="software-engineer"
    onClick={[Function]}
    type="checkbox"
    value="Software Engineer"
  />
  <label
    htmlFor="software-engineer"
    tabIndex={-1}
  >
    Software Engineer
  </label>
</div>
`;
